storage/innobase/trx/trx0sys.cc

```diff
@@ -408,122 +533,17 @@ void trx_sys_update_wsrep_checkpoint(
-        Same xid is re-persisted when TOI ends.
-        Latter call is still needed for DDL transaction that non-atomic.
-        So condition check is now >= and not just > */
-
-        /* Case 2:
-        We have the case when REPLACE INTO ... SELECT or
-        INSERT INTO ... SELECT is executed as TOI. This is done by
-        pt-table-checksum tool. In such case statement execution causes InnoDB
-        commit and storing of xid_seqno (wsrep_apply_cb) and then because of
-        TOI, storing of xid_seqno is requested again from wsrep_commit_cb with
-        the same xid_seqno.
-        Allow current and new values be the same, without introducing new flags
-        and logic to prevent double storing of the same value */
-
-        /* Case 3:
-
-           NBO allows interleaving of non-conflicting transactions. This allows
-           other transactions to be committed in InnoDB while there is an
-           ongoing NBO update. This is possible when the NBO thread has
-           finished phase one but is yet to start its phase two. Before adding
-           the below NBO condition, the other transactions updated its XID
-           there by resulting in the NBO thread to hit the below assertion as
-           it finds that the current thread no more has the latest xid in
-           InnoDB since NBO phase_two() calls this funcation with seqno X, the
-           one we got during phase one of the NBO. Please note that we do not
-           persist X+1+N+1 in SE upon completion on NBO even though it is only
-           incremented in galera.
-
-           i.e,
-           Seqno X       - Assigned during NBO begin
-           Seqno X+1     - Reserved for NBO end, but never used. Yes, this is a
-                           bug. TODO: Check why this happens.
-           Seqno X+1+N   - N Other transactions while NBO is in progress and
-                           the latest xid persisted in InnoDB
-           Seqno X+1+N+1 - Assigned during NBO end. It is incremented in
-                           galera, but never persisted in SE.
-                           TODO: Check why this happens.
-
-           As a result, the assertion evaluated to
-              "X" >= "X+1+N"
-        */
-        ut_ad((current_thd && wsrep_thd_is_in_nbo(current_thd)) ||
-              xid_seqno >= trx_sys_cur_xid_seqno);
-        trx_sys_cur_xid_seqno = xid_seqno;
-
-        /* Mark as done */
-        wsrep_unregister_from_group_commit(current_thd);
-      }
-    } else {
-      /* Wait and unregister from wsrep group commit */
-      wsrep_wait_for_turn_in_group_commit(current_thd);
+  if (wsrep_implicit_transaction(current_thd)) return;
 
-      memcpy(trx_sys_cur_xid_uuid, xid_uuid, 16);
-      trx_sys_cur_xid_seqno = xid_seqno;
-
-      /* Mark as done and unregister */
-      wsrep_unregister_from_group_commit(current_thd);
-    }
-  }
-
-  if (sys_header == NULL) {
-    sys_header = trx_sysf_get(mtr);
+  if (recovery && !should_store_recovery_xid(xid)) {
```

Comment:
From what the function name reads `should_store_recovery_xid`, it looks that we plan decide whether we need to restore the XID. But we are storing the values to `trx_sys_cur_xid_uuid` `trx_sys_cur_xid_seqno` instead. If storing is what we need, then can we instead rename the function name to `recover_xid()`?

Reply:
`should_store_recovery_xid` function does not store xid. xid is stored at the end of `trx_sys_update_wsrep_checkpoint`. Inside `should_store_recovery_xid` we only cache the current value for tracking purposes and decide if it should be stored or not.
Store == write to (sys_header + TRX_SYS_WSREP_XID_INFO)